src/base/bittorrent/customstorage.cpp

```diff
@@ -35,6 +35,191 @@
 #include "base/utils/fs.h"
 #include "common.h"
 
+#if (LIBTORRENT_VERSION_NUM >= 20000)
+#include <libtorrent/session.hpp>
+
+std::unique_ptr<lt::disk_interface> customDiskIOConstructor(
+        lt::io_context &ioContext, const lt::settings_interface &settings, lt::counters &counters)
+{
+    return std::make_unique<customDiskIOThread>(lt::default_disk_io_constructor(ioContext, settings, counters));
+}
+
+customDiskIOThread::customDiskIOThread(std::unique_ptr<libtorrent::disk_interface> nativeDiskIOThread)
+    : m_nativeDiskIO {std::move(nativeDiskIOThread)}
+{
+}
+
+lt::storage_holder customDiskIOThread::new_torrent(const lt::storage_params &storageParams, const std::shared_ptr<void> &torrent)
+{
+    lt::storage_holder storageHolder = m_nativeDiskIO->new_torrent(storageParams, torrent);
+
+    const QString savePath = Utils::Fs::expandPathAbs(QString::fromStdString(storageParams.path));
+    m_storageData[static_cast<lt::storage_index_t>(storageHolder)] = {
+            savePath
+            , storageParams.mapped_files ? *storageParams.mapped_files : storageParams.files
+            , storageParams.priorities};
+
+    return storageHolder;
+}
+
+void customDiskIOThread::remove_torrent(lt::storage_index_t storage)
+{
+    m_nativeDiskIO->remove_torrent(storage);
+}
+
+void customDiskIOThread::async_read(lt::storage_index_t storage, const lt::peer_request &peerRequest
+                                    , std::function<void (lt::disk_buffer_holder, const lt::storage_error &)> handler
+                                    , lt::disk_job_flags_t flags)
+{
+    m_nativeDiskIO->async_read(storage, peerRequest, handler, flags);
+}
+
+bool customDiskIOThread::async_write(lt::storage_index_t storage, const lt::peer_request &peerRequest
+                                     , const char *buf, std::shared_ptr<lt::disk_observer> diskObserver
+                                     , std::function<void (const lt::storage_error &)> handler, lt::disk_job_flags_t flags)
+{
+    return m_nativeDiskIO->async_write(storage, peerRequest, buf, diskObserver, handler, flags);
+}
+
+void customDiskIOThread::async_hash(lt::storage_index_t storage, lt::piece_index_t piece
+                                    , lt::span<lt::sha256_hash> hash, lt::disk_job_flags_t flags
+                                    , std::function<void (lt::piece_index_t, const lt::sha1_hash &, const lt::storage_error &)> handler)
+{
+    m_nativeDiskIO->async_hash(storage, piece, hash, flags, handler);
+}
+
+void customDiskIOThread::async_hash2(lt::storage_index_t storage, lt::piece_index_t piece
+                                     , int offset, lt::disk_job_flags_t flags
+                                     , std::function<void (lt::piece_index_t, const lt::sha256_hash &, const lt::storage_error &)> handler)
+{
+    m_nativeDiskIO->async_hash2(storage, piece, offset, flags, handler);
```

Comment:
in a lot of these calls you could move the `handler` into the next call, to save a copy. `std::move(handler)`